migrations/back_populate_uuids_plate_barcodes.py

```diff
@@ -0,0 +1,202 @@
+from crawler.sql_queries import SQL_MLWH_COUNT_MONGO_IDS, SQL_MLWH_GET_SAMPLE_FOR_MONGO_ID
+from crawler.types import Config, SampleDoc
+from migrations.helpers.shared_helper import (
+    extract_barcodes,
+    get_mongo_samples_for_source_plate,
+    mysql_generator,
+    validate_args,
+)
+
+LOGGER = logging.getLogger(__name__)
+
+FIELD_UUID_UPDATED = "uuid_updated"
+
+
+class ExceptionSampleWithSampleUUID(Exception):
+    pass
+
+
+class ExceptionSampleCountsForMongoAndMLWHNotMatching(Exception):
+    pass
+
+
+"""
+Iterates over the list of plate barcodes provided in a CSV file.
+Finds all the samples in MongoDB with that plate barcode and iterates over those.
+Looks for the same sample in the lighthouse_sample table using mongodb_id and checks whether it has an lh_sample_uuid.
+If not, the sample is skipped.
+Otherwise the lh_sample_uuid is added to the MongoDB document along with a key uuid_updated set to 'true'.
+"""
+
+
+def run(config: Config, s_filepath: str) -> None:
+    filepath = validate_args(config=config, s_filepath=s_filepath)
+
+    LOGGER.info("-" * 80)
+    LOGGER.info("STARTING BACK POPULATING UUIDS")
+    LOGGER.info(f"Time start: {datetime.now()}")
+
+    source_plate_barcodes = extract_barcodes(filepath=filepath)
+
+    LOGGER.info(f"Starting update process with input file {filepath}")
+
+    update_mongo_uuids(config=config, source_plate_barcodes=source_plate_barcodes)
+
+
+def update_mongo_uuids(config: Config, source_plate_barcodes: List[str]) -> None:
+    """Updates source plate and sample uuids in both mongo and mlwh
+
+    Arguments:
+        config {Config} -- application config specifying database details
+        source_plate_barcodes {List[str]} -- the list of source plate barcodes
+
+    Returns:
+        Nothing
+    """
+    # counters to track progress
+    counter_mongo_update_successes = 0
+    counter_mongo_update_failures = 0
+
+    with create_mongo_client(config) as client:
+        mongo_db = get_mongo_db(config, client)
+        samples_collection = get_mongo_collection(mongo_db, COLLECTION_SAMPLES)
+
+        check_samples_are_valid(config, samples_collection, source_plate_barcodes)
+
+        for source_plate_barcode in source_plate_barcodes:
+            LOGGER.info(f"Processing source plate barcode {source_plate_barcode}")
+
+            # List[SampleDoc]
+            sample_docs = get_mongo_samples_for_source_plate(samples_collection, source_plate_barcode)
+
+            # iterate through samples
+            for sample_doc in sample_docs:
+                # will every sample doc have a plate_barcode and lab id?
+                LOGGER.info(f"Sample in well {sample_doc[FIELD_COORDINATE]}")
+
+                # update sample in Mongo ‘samples’ to set lh_sample_uuid and updated_timestamp
+                try:
+                    query = SQL_MLWH_GET_SAMPLE_FOR_MONGO_ID % {MLWH_MONGODB_ID: sample_doc[FIELD_MONGODB_ID]}
```

Comment:
What if a sample from mongo doesnt have mongodb_id? This would query with NULL and returtn a lot of samples unrelated. In theory is not possible, but I haven't checked. Maybe we should log it? Reject? or raise error?

Reply:
I think every row in a mongodb collection has to have an id, it's fundamental to how it works?

Reply:
Yes this is the case.  MongoDB doesn't allow a missing `_id` field.  If a document is inserted without one, one gets generated automatically: https://www.mongodb.com/docs/manual/core/document/#:~:text=In%20MongoDB%2C%20each%20document%20stored%20in%20a%20collection,documents%20inserted%20through%20update%20operations%20with%20upsert%3A%20true.